Add some test cases for parsing/data_loader

alikins opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### COMPONENT NAME
core

##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.2.0 (dataloader_testcases 9bc5573b4b) last updated 2016/06/29 11:32:54 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 2f0f04437b) last updated 2016/06/29 10:48:47 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD d25c487ac8) last updated 2016/06/29 10:48:48 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY

Some potential unit test cases related to the dataload path_dwim changes.

Note some of these tests currently fail, either
because of unhandled exceptions on bad input (which
may be expected) or because the returned values
need to be verify correct.

So, this may just need tests updated, or it could
mean changes in the methods.

bcoca commented on test/units/parsing/test_data_loader.py:
```diff
@@ -62,6 +63,119 @@ def test_parse_fail_from_file(self, mock_def):
+        self.assertEquals(os.path.curdir, ret)
+
+    def test_all_slash(self):
+        ret = self._loader.path_dwim_relative('/', '/', '/')
```
dirname should never be a path, just a string

bcoca replied:
source is expected to be a file or path to a file, we should ALSO test directories, but the main focus should be files

bcoca commented on test/units/parsing/test_data_loader.py:
```diff
@@ -62,6 +63,119 @@ def test_parse_fail_from_file(self, mock_def):
+    def test_path_endswith_role_from_slash(self):
+        ret = self._loader.path_dwim_relative(path='foo/bar/tasks/', dirname='/', source='/')
+        assert ret is not None, "Returned path should not be None"
+        self.assertEquals(u'/', ret)
```
this would not be the expected return, it should be a list with one item, `foo/bar/tasks`

bcoca commented on test/units/parsing/test_data_loader.py:
```diff
@@ -62,6 +63,119 @@ def test_parse_fail_from_file(self, mock_def):
+        assert ret is not None, "Returned path should not be None"
+        self.assertEquals(u'/', ret)
+
+    def test_path_endswith_role(self):
```
a missing file should return None, but we should also test 'finding a file'

alikins committed:
more coverage

alikins committed:
add some test data to loader for coverage

alikins committed:
more tests, more coverage

alikins commented:
rebased and updated with more tests.

This also adds a few test data files for testing with real fs calls. Usually try to avoid that, but I think the extra test coverage is worth it here (and avoids lots of fragile/tedious mocking...)

alikins committed:
Fix type errors on ansible-vault script stdout

to_text() the popen.communicate() output so strip()
works, then back to_bytes().

py2 communicate() returns py2 native string while py3
returns bytes.

alikins committed:
vault.password is bytes, so fix type in text

alikins committed:
add missed units/parsing/data.yml to repo

alikins commented on test/units/parsing/data.yml:
```diff
@@ -0,0 +1,6 @@
+---
+- item1
```
Currently, the local data files needed for the tests is just added to the same directly as the test module.
A specific test data dir may be more appropriate (a fixtures/ subdir, or a test/units/data/ dir, etc) and can
be changed if this simpler approach isnt sufficient.

mattclay commented:
CI failure due to:

```
2017-02-01 23:26:28 =================================== FAILURES ===================================
2017-02-01 23:26:28 _______________________ TestDataLoader.test_is_directory _______________________
2017-02-01 23:26:28 
2017-02-01 23:26:28 self = <units.parsing.test_dataloader.TestDataLoader testMethod=test_is_directory>
2017-02-01 23:26:28 
2017-02-01 23:26:28     def test_is_directory(self):
2017-02-01 23:26:28         ret = self._loader.is_directory('parsing/test_data')
2017-02-01 23:26:28 >       self.assertTrue(ret)
2017-02-01 23:26:28 E       AssertionError: False is not true
2017-02-01 23:26:28 
2017-02-01 23:26:28 test/units/parsing/test_dataloader.py:88: AssertionError
```

mattclay commented:
Full logs for CI failure here: https://app.shippable.com/runs/58926c20d3cc6a1000c37e01/2/console

alikins committed:
re-re-re-remove loader.read_vault_password_file

alikins committed:
api updates

alikins committed:
remove bogus test ref'ing a relative path

mattclay commented:
@alikins I see the tests passed on the most recent commit. Is there anything else that needs to be done here, other than rebase the PR?

bcoca commented:
@alikins since i merged #23001 some of these tests might be obsolete or need to change
